charmstore/fakeclient.go

```diff
@@ -0,0 +1,582 @@
+	originalChannel := c.repo.channel
+	defer func() { c.repo.channel = originalChannel }()
+	return c.repo.WithChannel(channel).ListResources(id)
+}
+
+func (c FakeClient) GetResource(channel params.Channel, id *charm.URL, name string, revision int) (csclient.ResourceData, error) {
+	originalChannel := c.repo.channel
+	defer func() { c.repo.channel = originalChannel }()
+	return c.repo.WithChannel(channel).GetResource(id, name, revision)
+}
+
+func (c FakeClient) ResourceMeta(channel params.Channel, id *charm.URL, name string, revision int) (params.Resource, error) {
+	originalChannel := c.repo.channel
+	defer func() { c.repo.channel = originalChannel }()
+	return c.repo.WithChannel(channel).ResourceMeta(id, name, revision)
+}
+
+// ServerURL returns the empty string, as it is not accessing an actual charm store.
+//
+// ServerURL is part of the charmstore.csWrapper interface
+func (c FakeClient) ServerURL() string {
+	return "https://charmstore-testing.staging.jujucharms.com/"
+}
+
+func (c FakeClient) UploadCharm(id *charm.URL, charmData charm.Charm) (*charm.URL, error) {
+	return c.repo.UploadCharm(id, charmData)
+}
+
+func (c FakeClient) UploadCharmWithRevision(id *charm.URL, charmData charm.Charm, promulgatedRevision int) error {
+	return c.repo.UploadCharmWithRevision(id, charmData, promulgatedRevision)
+}
+
+func (c FakeClient) UploadBundle(id *charm.URL, bundleData charm.Bundle) (*charm.URL, error) {
+	return c.repo.UploadBundle(id, bundleData)
+}
+
+func (c FakeClient) UploadBundleWithRevision(id *charm.URL, bundleData charm.Bundle, promulgatedRevision int) error {
+	return c.repo.UploadBundleWithRevision(id, bundleData, promulgatedRevision)
+}
+
+func (c FakeClient) UploadResource(id *charm.URL, name, path string, file io.ReadSeeker, size int64, progress csclient.Progress) (revision int, err error) {
+	return c.repo.UploadResource(id, name, path, file, size, progress)
+}
+
+func (c FakeClient) CharmInfo(id string) (*charms.CharmInfo, error) {
+	url, err := charm.ParseURL(id)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	details, _, _, err := c.repo.ResolveWithChannel(url)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	info := &charms.CharmInfo{Revision: details.Revision, URL: id}
+	return info, nil
+}
+
+func (c FakeClient) AddLocalCharm(id *charm.URL, details charm.Charm, force bool) (*charm.URL, error) {
+	id = id.WithRevision(details.Revision())
+	return id, c.AddCharm(id, params.Channel("local"), force)
+}
+
+func (c FakeClient) AddCharm(id *charm.URL, channel params.Channel, force bool) error {
+	return c.repo.AddCharm(id, channel, force)
+}
+
+func (c FakeClient) AddCharmWithAuthorization(id *charm.URL, channel params.Channel, auth *macaroon.Macaroon, force bool) error {
+	return c.AddCharm(id, channel, force)
+}
+
+func (c FakeClient) AuthorizeCharmstoreEntity(*charm.URL) (*macaroon.Macaroon, error) {
+	return nil, nil
+}
+
+type ChannelAwareFakeClient struct {
```

Comment:
Some doc around this would be helpful.

Does the type need to be exported? I suppose it isn't to terrible given that this is all around testing.